packages/medusa/src/interfaces/services/event-bus.ts

```diff
@@ -1,6 +1,111 @@
-import { EntityManager } from "typeorm"
+import { ulid } from "ulid"
+import { StagedJob } from "../../models"
+import { TransactionBaseService } from "../transaction-base-service"
 
-export interface IEventBusService {
-  emit(event: string, data: any): Promise<void>
-  withTransaction(transactionManager?: EntityManager): this
+export type Subscriber<T = unknown> = (
+  data: T,
+  eventName: string
+) => Promise<void>
+
+export type SubscriberContext = {
+  subscriberId: string
+}
+
+export type SubscriberDescriptor = {
+  id: string
+  subscriber: Subscriber
+}
+
+export type EventHandler<T = unknown> = (
+  data: T,
+  eventName: string
+) => Promise<void>
+
+export interface IEventBusService extends TransactionBaseService {
+  emit<T>(
+    eventName: string,
+    data: T,
+    options?: unknown
+  ): Promise<StagedJob | void>
+
+  subscribe<T>(eventName: string, handler: EventHandler): this
+}
+
+export abstract class AbstractEventBusService
```

Comment:
self-review: Since `subscribe` and `unsubscribe` are likely to have identical business logic across modules and are not coupled with the messaging technology, I decided to create an abstract class.